controllers/cheuser/controller.go

```diff
@@ -0,0 +1,336 @@
+			} else {
+				return []reconcile.Request{}
+			}
+		}),
+	}
+}
+
+func isLabeledAsUserSettings(obj metav1.Object) bool {
+	return obj.GetLabels()["app.kubernetes.io/component"] == userSettingsComponentLabelValue
+}
+
+func (r *CheUserNamespaceReconciler) isInManagedNamespace(ctx context.Context, obj metav1.Object) bool {
+	info, err := r.namespaceCache.GetNamespaceInfo(ctx, obj.GetNamespace())
+	return err == nil && info != nil && info.OwnerUid != ""
+}
+
+func (r *CheUserNamespaceReconciler) triggerAllNamespaces(ctx context.Context) handler.EventHandler {
+	return &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(func(mo handler.MapObject) []reconcile.Request {
+			nss := r.namespaceCache.GetAllKnownNamespaces()
+			ret := make([]reconcile.Request, len(nss))
+
+			for _, ns := range nss {
+				ret = append(ret, reconcile.Request{
+					NamespacedName: types.NamespacedName{Name: ns},
+				})
+			}
+
+			return ret
+		}),
+	}
+}
+
+func (r *CheUserNamespaceReconciler) hasCheCluster(ctx context.Context, namespace string) bool {
+	list := v1.CheClusterList{}
+	if err := r.client.List(ctx, &list, client.InNamespace(namespace)); err != nil {
+		return false
+	}
+
+	return len(list.Items) > 0
+}
+
+func asReconcileRequestForNamespace(obj metav1.Object) []reconcile.Request {
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{Name: obj.GetNamespace()},
+		},
+	}
+}
+
+func (r *CheUserNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	info, err := r.namespaceCache.ExamineNamespace(ctx, req.Name)
+	if err != nil || info == nil || info.OwnerUid == "" {
+		return ctrl.Result{}, err
+	}
+
+	checluster := findManagingCheCluster(*info.CheCluster)
+	if checluster == nil {
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	if devworkspace.GetDevworkspaceState(r.scheme, checluster) != devworkspace.DevworkspaceStateEnabled {
+		return ctrl.Result{}, nil
+	}
+
+	// let's construct the deployContext to be able to use methods from v1 operator
+	deployContext := &deploy.DeployContext{
+		CheCluster: org.AsV1(checluster),
+		ClusterAPI: deploy.ClusterAPI{
+			Client:          r.client,
+			NonCachedClient: r.client,
+			DiscoveryClient: nil,
+			Scheme:          r.scheme,
+		},
+	}
+
+	isSelfSignedCertUsed, err := deploy.IsSelfSignedCertificateUsed(deployContext)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if isSelfSignedCertUsed {
+		if err = r.reconcileSelfSignedCert(ctx, deployContext, req.Name, checluster); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if err = r.reconcileProxySettings(ctx, req.Name, checluster, deployContext); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func findManagingCheCluster(key types.NamespacedName) *v2alpha1.CheCluster {
+	instances := devworkspace.GetCurrentCheClusterInstances()
+	if len(instances) == 0 {
+		return nil
+	}
+
+	if len(instances) == 1 {
```

Comment:
```suggestion
	if len(instances) > 1 {
```

Reply:
Nope, the logic is this:

If there is a single instance of `CheCluster` in the k8s cluster, then we don't need to look for it using the labels on the namespace (the `key`). If there are more than `CheCluster` instances, we're only able to figure out which one is in charge of this user's namespace by looking at the labels of the namespace.